lib/task_helpers.rb

```diff
@@ -1,122 +1,134 @@
-        # we catch up once before the alias move and once afterwards.
-        first_catchup_start_time = Time.now
-        adjusted_start_time = initial_start_time - (extra_catchup_minutes * 60)
-        catchup_index(adjusted_start_time, index_name, batch_size, sleep_time)
+      adjusted_start_time = initial_start_time - extra_catchup_minutes * 60
+      catchup_indices(index_names, adjusted_start_time, batch_size)
+      add_temporary_index_names(index_names)
+      LOG.info "Rebuild indices complete."
+    end
 
-        move_alias(alias_name, index_name, force_delete: true)
-        adjusted_start_time = first_catchup_start_time - (extra_catchup_minutes * 60)
-        catchup_index(adjusted_start_time, alias_name, batch_size, sleep_time)
+    # Get index name which corresponds to the model index name
+    def self.get_index_model_rel(model_index_name, current_indices)
+      current_indices.each do |index|
+        index_name = index.delete('^\[a-z]_')[0...-1]
+        if model_index_name == index_name
+          return index
+        end
       end
+    end
 
-      LOG.info "Rebuild index complete."
-      index_name
+    # Get model object which corresponds to the index name
+    def self.get_model_index_rel(current_index)
+      index_name = current_index.delete('^\[a-z]_')[0...-1]
+      INDEX_MODELS.each do |model|
+        if model.index_name == index_name
+          return model
+        end
+      end
     end
 
-    def self.catchup_index(start_time, index_name, batch_size=100, sleep_time=0)
-      [Comment, CommentThread].each do |model|
+    def self.catchup_indices(index_names, start_time, batch_size=100)
+      INDEX_MODELS.each do |model|
         current_batch = 1
-        model.where(:updated_at.gte => start_time).import(index: index_name, batch_size: batch_size) do |response|
-            batch_import_post_process(response, current_batch, sleep_time)
+        model.where(:updated_at.gte => start_time).import(index: model.index_name, batch_size: batch_size) do |response|
+            batch_import_post_process(response, current_batch)
             current_batch += 1
         end
       end
       LOG.info "Catch up from #{start_time} complete."
     end
 
-    def self.create_index(name=nil)
-      name ||= "#{Content::ES_INDEX_NAME}_#{Time.now.strftime('%Y%m%d%H%M%S%L')}"
-
-      Elasticsearch::Model.client.indices.create(index: name)
-      put_mappings(name)
-
-      LOG.info "Created new index: #{name}."
-      name
+    def self.create_indices
+      index_names = []
+      time_now = Time.now.strftime('%Y%m%d%H%M%S%L')
+
+      INDEX_MODELS.each do |model|
+        index_name = "#{model.index_name}_#{time_now}"
+        index_names.push(index_name)
+        Elasticsearch::Model.client.indices.create(
+          index: index_name,
+          body: {"mappings": model.mapping.to_hash}
+        )
+        move_alias(model.index_name, index_name, force_delete: true)
```

Comment:
I'm confused about some of the changes I see here - in the old code, we called `move_alias` at the end of `rebuild_index` so that, once the new index was ready, we moved the alias to point at it so that there'd be no downtime.

In the new code, we're calling `move_alias` in `create_indices`, which is in turn called at the _beginning_ of `rebuild_indices`... doesn't this mean we've moved the alias to point at an empty index, since we haven't yet populated it?  

I'd expect that we just leave the overall structure of all these functions roughly "the same" and modify them to work with our two indices instead of the single index we had before.  We know they were working already; it seems like we're incurring additional risk by making significant changes to the order/shape of the code.

Reply:
Sorry for waiting, I'm returning move alias like it was. Now I understand your question and I'm working on it. I think in 1 or 2 days it will be done